src/target/riscv/riscv-013.c

```diff
@@ -874,6 +879,12 @@ static int wait_for_idle(struct target *target, uint32_t *abstractcs)
 	assert(target);
 	assert(abstractcs);
 
+	dm013_info_t *dm = get_dm(target);
+	if (!dm) {
+		*abstractcs = 0;
+		return ERROR_FAIL;
+	}
```

Comment:
Can this situation happen in real life, or would that be a programmer error to call wait_for_idle() before the target is assigned to a DM?

If it is the latter, I'd recommend to add `LOG_ERROR("BUG: ...)` or use `assert(dm)` instead.

```suggestion
	if (!dm) {
		*abstractcs = 0;
		LOG_ERROR("BUG: Target %s is not assigned to any RISC-V debug module", target_name(target));
		return ERROR_FAIL;
	}
```

Reply:
Addressed. Used `LOG_ERROR("BUG:...)".

I have decided against an `assert()` since `wait_for_idle()` is quite often called in error handling, so it would be easy to miss something like:

```
int foo()
{
  if (bar() != ERROR_OK) /* bar() executes a command after calling get_dm() -> returns ERROR_FAIL if get_dm() returned NULL */
    goto error;
    ...
 error:
    wait_for_idle();
    return ERROR_FAIL;
}
```

In general, I don't like the fact that `get_dm()` allocates a new DM and can return NULL. I would like for allocation of a DM to be an explicit step in the examination process and for `get_dm()` to be a simple field accessor.

AFAIU, the only thing preventing such approach is the authentication mechanism -- the only way to authenticate a DM is to attempt an examination (to know the version of the spec used) and then to re-attempt it after issuing appropriate `riscv auth write/read` commands (here a DM is needed on a not-yet-examined target).

I think this mechanism should be re-imagined.

I will create a patch with a proposal, though I'm not sure when exactly :).